File: 2022/day_05.nim
include aoc

let
  raw = fetch(2022, 5).split("\n\n").mapIt(it.splitLines)

  stacks = collect(for x in 0..8:
    collect(for y in 0..7:
      if raw[0][7 - y][1 + 4 * x] != ' ': raw[0][7 - y][1 + 4 * x]))

  commands = collect(for line in raw[1]:
    line.scanTuple("move $i from $i to $i"))

var mut_stacks: seq[seq[char]]
proc stack(i: int): var seq[char] = mut_stacks[i - 1]
proc stringify: string = collect(for s in mut_stacks: s[^1]).join("")

part 1:
  mut_stacks = stacks
  for (_, a, b, c) in commands:
    for _ in 0..<a:
      c.stack.add b.stack.pop
  stringify()

part 2:
  mut_stacks = stacks
  for (_, a, b, c) in commands:
    c.stack.add b.stack.toOpenArray(b.stack.len - a, b.stack.high)
    b.stack.setLen b.stack.len - a
  stringify()

File: 2022/day_05.py
import aoc_lube
from aoc_lube.utils import extract_ints, chunk

def parse_raw():
    stacks, commands = aoc_lube.fetch(year=2022, day=5).split("\n\n")

    stacks = stacks.splitlines()
    stacks = [
        [stacks[y][x] for y in range(7, -1, -1) if stacks[y][x] != " "]
        for x in range(1, 35, 4)
    ]

    return stacks, tuple(chunk(extract_ints(commands), 3))

STACKS, COMMANDS = parse_raw()

def part_one():
    for a, b, c in COMMANDS:
        for _ in range(a):
            STACKS[c - 1].append(STACKS[b - 1].pop())

    return "".join(map(list.pop, STACKS))

def part_two():
    for a, b, c in COMMANDS:
        STACKS[c - 1].extend(STACKS[b - 1][-a:])
        del STACKS[b - 1][-a:]

    return "".join(map(list.pop, STACKS))

aoc_lube.submit(year=2022, day=5, part=1, solution=part_one)
aoc_lube.submit(year=2022, day=5, part=2, solution=part_two)
